File: 프로그래머스/1/67256. ［카카오 인턴］ 키패드 누르기/［카카오 인턴］ 키패드 누르기.java
import java.util.*;

class Solution {
    
    private static final Map<String, int[]> keypad = new HashMap<>();

    private String calculateDistance(String preLeft, String preRight, int number, String hand) {
        
        // 위치 정보 넣기
        keypad.put("1", new int[]{0, 0}); 
        keypad.put("2", new int[]{0, 1});
        keypad.put("3", new int[]{0, 2});
        keypad.put("4", new int[]{1, 0}); 
        keypad.put("5", new int[]{1, 1}); 
        keypad.put("6", new int[]{1, 2});
        keypad.put("7", new int[]{2, 0}); 
        keypad.put("8", new int[]{2, 1}); 
        keypad.put("9", new int[]{2, 2});
        keypad.put("*", new int[]{3, 0}); 
        keypad.put("0", new int[]{3, 1}); 
        keypad.put("#", new int[]{3, 2});
        
        int[] L_pos = keypad.get(preLeft);
        int[] R_pos = keypad.get(preRight);
        int[] N_pos = keypad.get(Integer.toString(number));

        // 맨해튼 거리 계산: |x1 - x2| + |y1 - y2|
        int distL = Math.abs(L_pos[0] - N_pos[0]) + Math.abs(L_pos[1] - N_pos[1]);
        int distR = Math.abs(R_pos[0] - N_pos[0]) + Math.abs(R_pos[1] - N_pos[1]);

        if (distL < distR) {
            return "L";
        } else if (distR < distL) {
            return "R";
        } else {
            // 거리가 같을 경우 주 손잡이 확인
            return hand.equals("left") ? "L" : "R";
        }
    }
    
    
    public String solution(int[] numbers, String hand) {
        StringBuilder sb = new StringBuilder();
        
        String preLeft = "*";
        String preRight = "#";
        
        for(int number : numbers){
            if(number == 1 || number == 4 || number == 7){
                preLeft = Integer.toString(number);
                sb.append("L");
            }else if(number == 3 || number == 6 || number == 9){
                preRight = Integer.toString(number);
                sb.append("R");
            }else{
                String result = calculateDistance(preLeft, preRight, number, hand);
                if (result.equals("R")) {
                    preRight = Integer.toString(number);
                    sb.append("R");
                } else {
                    preLeft = Integer.toString(number);
                    sb.append("L");
                }
            }
        }
        
        return sb.toString();
    }
}

File: 프로그래머스/1/67256. ［카카오 인턴］ 키패드 누르기/［카카오 인턴］ 키패드 누르기.py
def calculate_distance(pre_left, pre_right, number,hand):
    # 키패드 위치 좌표 사전
    keypad = {
        1: (0, 0), 2: (0, 1), 3: (0, 2),
        4: (1, 0), 5: (1, 1), 6: (1, 2),
        7: (2, 0), 8: (2, 1), 9: (2, 2),
        '*': (3, 0), 0: (3, 1), '#': (3, 2)
    }
    
    # 각 위치의 좌표 가져오기
    L_pos = keypad[pre_left]
    R_pos = keypad[pre_right]
    N_pos = keypad[number]
    
    # 맨해튼 거리 계산
    dist_l = abs(L_pos[0] - N_pos[0]) + abs(L_pos[1] - N_pos[1])
    dist_r = abs(R_pos[0] - N_pos[0]) + abs(R_pos[1] - N_pos[1])
    
    if dist_l < dist_r:
        return 'L'
    elif dist_r < dist_l:
        return 'R'
    else:
        # 거리가 같으면 주 손잡이 기준
        return 'L' if hand == 'left' else 'R'
    

def solution(numbers, hand):
    answer = ''
    
    pre_right = '*' # 오른손이 최근에 누른 버튼의 번호
    pre_left = '#' # 왼손이 최근에 누른 버튼의 번호
    
    for number in numbers:
        if number in [1,4,7]:
            pre_left = number
            answer += 'L'
        elif number in [3,6,9]:
            pre_right = number
            answer += 'R'
        else:
            if calculate_distance(pre_left, pre_right, number,hand) == 'R':
                pre_right = number
                answer += 'R'
            else:
                pre_left = number
                answer += 'L'
    
    return answer